build/provenance.go

```diff
@@ -0,0 +1,198 @@
+		bo := backoff.NewExponentialBackOff()
+		bo.MaxElapsedTime = 5 * time.Second
+		return backoff.Retry(func() error {
+			res, err := fetchProvenance(ctx, c, ref, mode, l)
+			if err != nil {
+				return err
+			}
+			for k, v := range res {
+				sr.ExporterResponse[k] = v
+			}
+			return nil
+		}, bo)
+	})
+}
+
+func fetchProvenance(ctx context.Context, c *client.Client, ref string, mode confutil.MetadataProvenanceMode, l progress.SubLogger) (out map[string]string, err error) {
+	cl, err := c.ControlClient().ListenBuildHistory(ctx, &controlapi.BuildHistoryRequest{
+		Ref:       ref,
+		EarlyExit: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	var mu sync.Mutex
+	eg, ctx := errgroup.WithContext(ctx)
+	for {
+		ev, err := cl.Recv()
+		if errors.Is(err, io.EOF) {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+		if ev.Record == nil {
+			continue
+		}
+		if ev.Record.Result != nil {
+			provenanceDgst := provenanceDigest(ev.Record.Result)
+			if provenanceDgst == nil {
+				continue
+			}
+			eg.Go(func() error {
+				dt, err := getBlob(ctx, c, *provenanceDgst, "", l)
+				if err != nil {
+					return errors.Wrapf(err, "failed to load provenance from build record")
+				}
+				prv, err := encodeProvenance(dt, mode)
+				if err != nil {
+					return err
+				}
+				mu.Lock()
+				if out == nil {
+					out = make(map[string]string)
+				}
+				out["buildx.build.provenance"] = prv
+				mu.Unlock()
+				return nil
+			})
+		} else if ev.Record.Results != nil {
+			for platform, res := range ev.Record.Results {
+				platform := platform
+				provenanceDgst := provenanceDigest(res)
+				if provenanceDgst == nil {
+					continue
+				}
+				eg.Go(func() error {
+					dt, err := getBlob(ctx, c, *provenanceDgst, platform, l)
+					if err != nil {
+						return errors.Wrapf(err, "failed to load provenance from build record")
+					}
+					prv, err := encodeProvenance(dt, mode)
+					if err != nil {
+						return err
+					}
+					mu.Lock()
+					if out == nil {
+						out = make(map[string]string)
+					}
+					out["buildx.build.provenance/"+platform] = prv
+					mu.Unlock()
+					return nil
+				})
+			}
+		}
+	}
+	return out, eg.Wait()
+}
+
+func provenanceDigest(res *controlapi.BuildResultInfo) *digest.Digest {
+	for _, a := range res.Attestations {
+		if a.MediaType == "application/vnd.in-toto+json" && strings.HasPrefix(a.Annotations["in-toto.io/predicate-type"], "https://slsa.dev/provenance/") {
+			return &a.Digest
+		}
+	}
+	return nil
+}
+
+func encodeProvenance(dt []byte, mode confutil.MetadataProvenanceMode) (string, error) {
+	if mode == confutil.MetadataProvenanceModeMax {
+		return base64.StdEncoding.EncodeToString(dt), nil
+	}
+
+	var prv provenancetypes.ProvenancePredicate
+	if err := json.Unmarshal(dt, &prv); err != nil {
+		return "", errors.Wrapf(err, "failed to unmarshal provenance")
+	}
+
+	// remove buildConfig and metadata for minimal provenance
```

Comment:
Should be reset the whole invocation block as well. Atm looks some partial info that I don't think would be very useful.

The builder ID shows up as empty string as well. I think we can mark it omitempty in buildkit (we could also wrap the type in here and mark it omitempty here).

Reply:
Hum looks like it's not marked as omitempty in https://github.com/in-toto/in-toto-golang/blob/2ef7537b66dab4620b440536d3065e5366701f01/in_toto/slsa_provenance/common/common.go#L9. On purpose?

Reply:
Now wrapping directly in buildx as it seems fine for the metadata file case but don't think we should omitempty in BuildKit related to https://github.com/docker/buildx/pull/2280#discussion_r1559113780